File: agent/agent_config.py
import os
import configparser
from typing import Dict, List, Optional, Iterator

class AgentConfig:
    def __init__(self, agent_name: str = None):
        self.options = configparser.ConfigParser()
        self.options.add_section('DEFAULT')
        if agent_name:
            self.set_agent_name(agent_name)
    
    @classmethod
    def from_file(cls, path: str) -> 'AgentConfig':
        if not os.path.exists(path):
            raise FileNotFoundError(f"Config file not found: {path}")
        
        config = cls()
        config.options.read(path)
        return config
    
    @classmethod
    def from_stream(cls, input_stream) -> 'AgentConfig':
        config = cls()
        config.options.read_file(input_stream)
        return config
    
    def get_agent_name(self) -> str:
        return self.get_option('agent_name')
    
    def set_agent_name(self, agent_name: str) -> 'AgentConfig':
        self.set_option('agent_name', agent_name)
        return self
    
    def get_broker_address(self) -> str:
        return self.get_option('broker_address')
    
    def set_broker_address(self, broker_address: str) -> 'AgentConfig':
        self.set_option('broker_address', broker_address)
        return self
    
    def get_place_name(self) -> str:
        return self.get_option('place_name')
    
    def set_place_name(self, place_name: str) -> 'AgentConfig':
        self.set_option('place_name', place_name)
        return self
    
    def get_option(self, option_name: str) -> Optional[str]:
        return self.options.get('DEFAULT', option_name, fallback=None)
    
    def get_option_or_default(self, option_name: str, default_value: str) -> str:
        return self.options.get('DEFAULT', option_name, fallback=default_value)
    
    def get_option_as_array(self, option_name: str) -> List[str]:
        value = self.get_option(option_name)
        if not value:
            return []
        return [item.strip() for item in value.split(',')]
    
    def set_option(self, option_name: str, option_value: str) -> 'AgentConfig':
        self.options.set('DEFAULT', option_name, option_value)
        return self
    
    def list_option_names(self) -> Iterator[str]:
        return self.options['DEFAULT'].keys()

File: lapras-middleware/src/main/java/kr/ac/kaist/cdsn/lapras/AgentConfig.java
package kr.ac.kaist.cdsn.lapras;

import org.apache.commons.configuration.ConfigurationException;
import org.apache.commons.configuration.PropertiesConfiguration;

import java.io.*;
import java.util.Iterator;

/**
 * Created by Daekeun Lee on 2016-11-14.
 */
public class AgentConfig {
    private final PropertiesConfiguration options = new PropertiesConfiguration();

    public AgentConfig(String agentName) {
        options.setProperty("agent_name", agentName);
    }

    private AgentConfig() {}

    public static AgentConfig fromFile(String path) throws FileNotFoundException {
        AgentConfig agentConfig = new AgentConfig();
        try {
            agentConfig.options.load(new FileReader(path));
        } catch (ConfigurationException e) {
            return null;
        }
        return agentConfig;
    }

    public static AgentConfig fromStream(InputStream inputStream) throws IOException {
        AgentConfig agentConfig = new AgentConfig();
        try {
            agentConfig.options.load(new InputStreamReader(inputStream));
        } catch (ConfigurationException e) {
            return null;
        }
        return agentConfig;
    }

    public String getAgentName() {
        return getOption("agent_name");
    }

    public AgentConfig setAgentName(String agentName) {
        setOption("agent_name", agentName);
        return this;
    }

    public String getBrokerAddress() {
        return getOption("broker_address");
    }

    public AgentConfig setBrokerAddress(String brokerAddress) {
        setOption("broker_address", brokerAddress);
        return this;
    }

    public String getPlaceName() {
        return getOption("place_name");
    }

    public AgentConfig setPlaceName(String placeName) {
        setOption("place_name", placeName);
        return this;
    }

    public String getOption(String optionName) {
        return options.getString(optionName);
    }

    public String getOptionOrDefault(String optionName, String defaultValue) {
        String option = getOption(optionName);
        if(option != null) return option;
        return defaultValue;
    }

    public String[] getOptionAsArray(String optionName) {
        return options.getStringArray(optionName);
    }

    public AgentConfig setOption(String optionName, String optionValue) {
        options.setProperty(optionName, optionValue);
        return this;
    }

    public Iterator<String> listOptionNames() {
        return options.getKeys();
    }
}
